internal/k8s/controllers/config_conversion_test.go

```diff
@@ -0,0 +1,235 @@
+			},
+			Spec: v1beta1.IPAddressPoolSpec{
+				Addresses: []string{
+					"10.10.0.0/16",
+				},
+				AllocateTo: &v1beta1.ServiceAllocation{
+					Namespaces: []string{"foo", "bar"},
+				},
+			},
+		},
+	}
+	bgpAdvs := []v1beta1.BGPAdvertisement{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "adv1",
+				Namespace: "metallb-system",
+			},
+			Spec: v1beta1.BGPAdvertisementSpec{
+				Communities: []string{"bar"},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "adv2",
+				Namespace: "metallb-system",
+			},
+			Spec: v1beta1.BGPAdvertisementSpec{
+				Communities: []string{"bar2"},
+			},
+		},
+	}
+	l2Advs := []v1beta1.L2Advertisement{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "l2adv1",
+				Namespace: "metallb-system",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "l2adv2",
+				Namespace: "metallb-system",
+			},
+			Spec: v1beta1.L2AdvertisementSpec{
+				Interfaces: []string{"foo"},
+			},
+		},
+	}
+
+	communities := []v1beta1.Community{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "community",
+				Namespace: "metallb-system",
+			},
+			Spec: v1beta1.CommunitySpec{
+				Communities: []v1beta1.CommunityAlias{
+					{
+						Name:  "bar",
+						Value: "64512:1234",
+					},
+				},
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "community1",
+				Namespace: "metallb-system",
+			},
+			Spec: v1beta1.CommunitySpec{
+				Communities: []v1beta1.CommunityAlias{
+					{
+						Name:  "bar2",
+						Value: "64512:1235",
+					},
+				},
+			},
+		},
+	}
+	namespaces := []v1.Namespace{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foo",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "bar",
+			},
+		},
+	}
+	nodes := []v1.Node{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node2",
+			},
+		},
+	}
+
+	oldConfig, err := toConfig(config.DontValidate,
+		pools,
+		peers,
+		bfdProfiles,
+		l2Advs,
+		bgpAdvs,
+		[]v1beta1.AddressPool{},
+		communities,
+		map[string]v1.Secret{},
+		nodes,
+		namespaces,
+		v1.ConfigMap{})
+
+	if err != nil {
+		t.Fatalf("conversion failed")
+	}
+	seed := time.Now().UnixNano()
+	rand.Seed(seed)
+
+	for i := 0; i < 100; i++ {
+		shuffleObjects(peers)
+		shuffleObjects(bfdProfiles)
+		shuffleObjects(pools)
+		shuffleObjects(bgpAdvs)
+		shuffleObjects(l2Advs)
+		shuffleObjects(communities)
+		shuffleObjects(nodes)
+		shuffleObjects(namespaces)
+
+		config, err := toConfig(config.DontValidate,
+			pools,
+			peers,
+			bfdProfiles,
+			l2Advs,
+			bgpAdvs,
+			[]v1beta1.AddressPool{},
+			communities,
+			map[string]v1.Secret{},
+			nodes,
+			namespaces,
+			v1.ConfigMap{})
+
+		if err != nil {
+			t.Fatalf("conversion failed")
+		}
+
+		if !reflect.DeepEqual(config, oldConfig) {
+			t.Fatalf("conversion is not stable, seed %d, %v\n %v\n", seed, spew.Sdump(config), spew.Sdump(oldConfig))
```

Comment:
nit: let's add these small details so the output will be clearer
```suggestion
			t.Fatalf("conversion is not stable, seed %d, got: %v\n expected: %v\n", seed, spew.Sdump(config), spew.Sdump(oldConfig))
```

Reply:
changed